compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] =
+      if n[0].kind == nkEmpty: empty(n[0].info)
+      else:                    expr(c, n[0])
+
+  of nkFinally:
+    invalid()
+  of nkBlockStmt:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = stmt(c, n[1])
+  of nkStaticStmt:
+    checkSonsLen(n, 1)
+    result = stmt(c, n[0])
+  of nkStmtList:
+    result = stmts(c, n)
+  of nkImportStmt, nkImportExceptStmt, nkExportStmt, nkExportExceptStmt, nkFromStmt, nkIncludeStmt:
+    # QUESTION: should we perform any additional validation here or leave that
+    #           to sem?
+    checkMinSonsLen(n, 1)
+    result = exprs(c, n)
+  of nkBindStmt, nkMixinStmt:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] = qualifiedIdent(c, n[i])
+
+  of nkUsingStmt:
+    # we're much more strict than the parser here: everything besides an ident
+    # def is disallowed
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] =
+        case n.kind
+        of nkIdentDefs:   strictIdentDefs(c, n[i])
+        of nkCommentStmt: commentStmt(c, n[i])
+        else:             invalidAst(c, n[i])
+
+  of nkCommentStmt:
+    result = commentStmt(c, n)
+  of nkStmtListExpr:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = expr(c, n[^1])
+  of nkBlockExpr:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = expr(c, n[1])
+  of nkStmtListType:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = typeExpr(c, n[^1])
+  of nkBlockType:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = typeExpr(c, n[1])
+
+  of nkWith, nkWithout:
+    unreachable("missing")
+
+  of nkTypeOfExpr:
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+  of nkObjectTy,
+    nkTupleTy    ,
+    nkTupleClassTy,
+    nkTypeClassTy  ,
+    nkStaticTy:
+    unreachable("missing")
+  of nkRecList,
+    nkRecCase,
+    nkRecWhen:
+    # context dependent
+    invalid()
+  of nkRefTy,
+    nkPtrTy,
+    nkVarTy,
+    nkConstTy,
+    nkMutableTy,
+    nkDistinctTy,
+    nkProcTy,
+    nkIteratorTy,
+    nkSharedTy,
+    nkEnumTy,
+    nkEnumFieldDef:
+    invalid()
+  of nkArgList:
+    unreachable("missing")
+  of nkPattern:
+    # disallow in macro output
+    invalid()
+  of nkTryStmt, nkHiddenTryStmt:
+    # treat a hidden 'try' statement as a normal one
+    result = parseTry(c, n)
+  of nkClosure:
+    # can't be expressed in untyped AST. Also only present in post-transform
+    # AST
+    # XXX: we could throw away the 'env' part and only use the 'prc' -- I'm
+    #      not sure if that's a good idea however
+    invalid()
```

Comment:
For this PR, I think dropping the env var might not be the way to go, maybe we want to go with:

1. write out locally the:
    1. env type
    2. procdef
2. call it to simulate it

----

## my thinking

There are two paths here when it comes to defining what a "raisings" means:
1. go back in time
5. use "reverse" transforms that emit roughly equivalent untyped code

I think 1 is nice where we can reasonably rewind thing, but I don't believe this is a property we can preserve unless we keep a) a full history of transforms, b) forward transforms need to preserve things, c) that last point discounts fusing, and d) I don't think the user always wants a time machine.

I believe the correct answer is phase dependent, so from post-`transf` is likely too big a transform (`hlo` has run) and history rewind is likely what people do not want.  We can't handle this sort of decisioning right now. If it's early sem (macro/template expansion), back to untyped is better as a history rewind.

We don't track enough data to do this, hence my suggestion above for now.

Reply:
The problem is that there is no equivalent for `nkClosure` in the untyped AST. A `nkClosure` is essentially a `nkTupleConstr` that always has two elements and is always of `tyProc` type. Consider the following code:
```nim
proc aux(x: proc(p: int)) = discard

proc prc() =
  var x = 0
  var y: int
  
  proc inner(p: int) =
    x = 1
    y = p

  inner(1)       # 1
  aux(inner)     # 2
  let cl = inner # 3
  cl(1)
```
After `transf` the body of `prc` will look like this:
```nim
var :env
internalNew(:env)

var :env.x1 = 0
var :env.y2: int

proc inner(p: int; :envP) = ...

(inner, :env)(1)       # 1
# ^ `nkClosure`
aux((inner, :env))     # 2
#   ^ `nkClosure`
let cl = (inner, :env) # 3
#        ^ `nkClosure`
cl(1)
```

For 1 and 3 it's possible to turn the `nkClosure` back into untyped AST (knowledge of sibling nodes is required, however), bit for 2 it's not.

Right now, what a node unwinds to depends only on the node kind plus the context it is used in. Supporting unwinding the above transformed code would mean that what a node translates to is also dependent on the sub-nodes (and sometimes even sibling nodes).

Reply:
My terrible idea here was that we wouldn't do that much recovery. Instead we'd emit the creation of the env etc... but that runs into all sorts of issues, with `internalNew` and a bunch more stuff leaking out.

Keeping it invalid for now makes sense.

Reply:
Thanks for the detailed explanation with the before and after transformation.